2007-07-28  Ulrich Drepper  <drepper@redhat.com>

	[BZ 4858]
	* stdio-common/printf_fp.c (___printf_fp): Fix special case of
	#.0g and value rounded to 1.0.

===================================================================
RCS file: /cvs/glibc/libc/stdio-common/printf_fp.c,v
retrieving revision 1.63
retrieving revision 1.64
diff -u -r1.63 -r1.64
--- libc/stdio-common/printf_fp.c	2007/04/30 22:31:21	1.63
+++ libc/stdio-common/printf_fp.c	2007/07/29 00:25:19	1.64
@@ -986,7 +986,9 @@
 	    if (*wtp != decimalwc)
 	      /* Round up.  */
 	      (*wtp)++;
-	    else if (__builtin_expect (spec == 'g' && type == 'f' && info->alt,
+	    else if (__builtin_expect (spec == 'g' && type == 'f' && info->alt
+				       && wtp == wstartp + 1
+				       && wstartp[0] == L'0',
 				       0))
 	      /* This is a special case: the rounded number is 1.0,
 		 the format is 'g' or 'G', and the alternative format
